Entity/RewardProperties: Account Tier based Reward Properties. Retail-like.

kirmmin opened the pull request:
This provides the necessary functionality to send unlocks to the client based on their Account Tier. The Account Tier itself is unimportant for the purpose of this feature. My goal was to make sure that we were sending things like Guild/Circle functionality, Bag Slots, Bank Slots, Auction and Commodity Counts appropriately. The Game Tables have this data available in it, so I am using them to generate the necessary properties.
In addition, the EntitlementManager can now be used to update the Client immediately. When I "purchase" a Costume slot, it is immediately reflected in my Character Pane without relog/zone. 

This was necessary work to get Store Purchasing working. This PR, #272, #273, and #275, are all necessary for Store Purchases to work. Outside of the added functionality, with the Store Purchasing, the majority of "Account features" should be complete.

Rawaho commented on Source/NexusForever.WorldServer/Network/Message/Model/ServerAccountTier.cs:
```diff
@@ -0,0 +1,19 @@
+using System.Collections.Generic;
```
Remove unneeded includes

Rawaho commented on Source/NexusForever.WorldServer/Game/Entity/Player.cs:
```diff
@@ -153,6 +153,12 @@ public uint VehicleGuid
+        /// All <see cref="RewardProperty"/> that are sent to the Player.
+        /// </summary>
+        /// <remarks>This is keyed by <see cref="RewardPropertyType"/> and Data key. This Data key is usually a "type", e.g. (RewardPropertyType.AccountCurrency, 15) would be Crimson Essence. This is used to determine bonuses that are specific to a type of RewardProperty.</remarks>
+        public Dictionary<(RewardPropertyType, uint), RewardProperty> Rewards { get; } = new Dictionary<(RewardPropertyType, uint), RewardProperty>();
```
Should really move this and any methods related to it to a manager, this could be a new one or the existing entitlement manager since they seem related.

Rawaho commented on Source/NexusForever.WorldServer/Game/Entity/Player.cs:
```diff
@@ -782,6 +757,18 @@ public void SendSystemMessage(string text)
             });
         }
 
+        private void CalculateRewardProperties()
```
See above

Rawaho commented on Source/NexusForever.WorldServer/Game/AssetManager.cs:
```diff
@@ -157,6 +162,23 @@ private void CacheCreatureTargetGroups()
+            for (int i = 0; i < rewardProps.Count(); i++)
+                entries.Add((AccountTier)i, new List<RewardPropertyPremiumModifierEntry>());
+
+            foreach (var rewardPropTier in rewardProps)
```
Could clean this nested foreach with maybe SelectMany?

Rawaho commented on Source/NexusForever.WorldServer/Game/AssetManager.cs:
```diff
@@ -157,6 +162,23 @@ private void CacheCreatureTargetGroups()
+            var entries = ImmutableDictionary.CreateBuilder<AccountTier, List<RewardPropertyPremiumModifierEntry>>();
+
+            // Unknown how to set the PremiumSystemEnum in the Client at this time. In addition, looks like only "Hybrid" was used on Retail.
+            var rewardProps = GameTableManager.Instance.RewardPropertyPremiumModifier.Entries.Where(i => (PremiumSystem)i.PremiumSystemEnum == PremiumSystem.Hybrid).GroupBy(i => i.Tier);
```
With long nested LINQ statements try and have each new call on a new line.
```
GameTableManager.Instance.RewardPropertyPremiumModifier.Entries
	.Where(i => (PremiumSystem)i.PremiumSystemEnum == PremiumSystem.Hybrid)
	.GroupBy(i => i.Tier);
```

Rawaho commented on Source/NexusForever.WorldServer/Game/AssetManager.cs:
```diff
@@ -157,6 +162,23 @@ private void CacheCreatureTargetGroups()
+                    foreach ((AccountTier tier, List<RewardPropertyPremiumModifierEntry> list) in entries.Where(e => e.Key >= (AccountTier)entry.Tier)) // Base Reward Properties are determined by current Account Tier and lower. This is how Retail worked.
+                        list.Add(entry);
+
+            rewardPropertiesByTier = entries.ToImmutableDictionary(e => e.Key, e=> e.Value.ToImmutableList());
```
I believe you want ToImmutable rather than ToImmutableDictionary for entries?

kirmmin replied:
For this, I did it the same way as the one above because I wanted to build an `ImmutableDictionary` and create an `ImmutableList` at the same time. I can see if `ToImmutable` allows the same overload, though.

kirmmin replied:
It doesn't. I can refacotr, but it will be more lines of code and  could be more confusing. It might make sense to leave it as it is to keep it as clean as possible?

Rawaho commented on Source/NexusForever.WorldServer/Game/Entity/EntitlementManager.cs:
```diff
@@ -165,5 +170,22 @@ private static T SetEntitlement<T>(IDictionary<EntitlementType, T> collection, E
+
+            foreach (RewardPropertyPremiumModifierEntry rewardPropertyEntry in GameTableManager.Instance.RewardPropertyPremiumModifier.Entries.Where(i => i.Tier <= (uint)session.AccountTier && (EntitlementType)i.EntitlementIdModifierCount == type))
+            {
+                if (!session.Player.Rewards.TryGetValue(((RewardPropertyType)rewardPropertyEntry.RewardPropertyId, rewardPropertyEntry.RewardPropertyData), out RewardProperty rewardProperty))
```
Try and reduce direct public access to a collection like this, encapsulate with methods.

kirmmin commented:
Refactored following the Review. I moved the methods to the `EntitlementManager` per your suggestion and made sure I encapsulated with methods. Not sure why I didn't do it prior to submitting the PR other than just overlooked it. I left one of the conversations outstanding because I didn't want to resolve until you've answered.
